mirror/modules/source/haproxy_spoe.go

```diff
@@ -146,6 +150,103 @@ func (m *HAProxySPOE) handleMessage(args []spoe.Message) ([]spoe.Action, error)
 	return nil, nil
 }
 
-func logBadMessage(msg string) {
-	log.Errorf("%s: bad message: %s", HAProxySPOEName, msg)
+func mapMethod(req *mirror.Request, value interface{}) error {
+	method, ok := value.(string)
+	if !ok {
+		return fmt.Errorf("bad type %T received for method, expected string", value)
+	}
+
+	i, ok := mirror.Method_value[method]
+	if !ok {
+		return fmt.Errorf("unknown method %q", value)
+	}
+
+	req.Method = mirror.Method(i)
+	return nil
+}
+
+func mapPath(req *mirror.Request, value interface{}) error {
+	path, ok := value.(string)
+	if !ok {
+		return fmt.Errorf("bad type %T received for method, expected string", value)
+	}
+
+	req.Path = path
+	return nil
+}
+
+func mapVer(req *mirror.Request, value interface{}) error {
+	ver, ok := value.(string)
+	if !ok {
+		return fmt.Errorf("bad type %T received for method, expected string", value)
+	}
+
+	switch ver {
+	case "1.1":
+		req.HttpVersion = mirror.HTTPVersion_HTTP1_1
+	case "2":
+		req.HttpVersion = mirror.HTTPVersion_HTTP2
+	default:
+		req.HttpVersion = mirror.HTTPVersion_HTTP1_0
+	}
+
+	return nil
+}
+
+func mapHeaders(req *mirror.Request, value interface{}) error {
+	rawHeaders, ok := value.([]byte)
+	if !ok {
+		return fmt.Errorf("bad type %T received for headers, expected bytes", value)
+	}
+
+	headers, err := spoe.DecodeHeaders(rawHeaders)
+	if err != nil {
+		return err
+	}
+
+	req.Headers = map[string]*mirror.HeaderValue{}
+	for name, values := range headers {
+		req.Headers[name] = &mirror.HeaderValue{Values: values}
+	}
+
+	return nil
```

Comment:
Why return Nil in mapping function? Aren't we suppose to return the actual mapping value?

Reply:
No, in these mapping functions we are modifying the request given as argument (notice the pointer). As pointed out by the method signature we are only returning an error if something wrong happened.